File: library_management/library_management/report/library_book_dropdown_report2/library_book_dropdown_report2.py
import frappe
def execute(filters=None):
    columns = [
        {"label": "Book Name", "fieldname": "book_name", "fieldtype": "Data", "width": 200},
        {"label": "Author", "fieldname": "author", "fieldtype": "Data", "width": 200},
        {"label": "Email", "fieldname": "email", "fieldtype": "Data", "width": 200},
        {"label": "Action", "fieldname": "action", "fieldtype": "HTML", "width": 150},
    ]
    data = []
    for book in frappe.get_all("Library Book", fields=["name", "book_name", "author", "email"]):
        data.append({
            "book_name": book.book_name,
            "author": book.author,
            "email": book.email,
            "action": f"""
                <select class='book-action' data-name='{book.name}'>
                    <option value="">Select</option>
                    <option value="Available">Available</option>
                    <option value="Issued">Issued</option>
                    <option value="Lost">Lost</option>
                </select>
            """
        })
    return columns, data
@frappe.whitelist()
def update_action(docname, action):
    frappe.db.set_value("Library Book", docname, "status", action)
    frappe.db.commit()

File: library_management/library_management/report/library_book_dropdown_report2/library_book_dropdown_report2.js
// Copyright (c) 2025, shridevi and contributors
// For license information, please see license.txt

frappe.query_reports["Library Book Dropdown Report2"] = {
    onload: function(report) {
        report.page.set_title("Library Book Dropdown Report2");

        // Wait for the report to render
        frappe.after_ajax(() => {
            // Use event delegation to handle dropdown changes
            $(document).on("change", ".book-action", function() {
                const book_name = $(this).data("name");
                const selected_value = $(this).val();

                frappe.call({
                    method: "library_management.library_management.report.library_book_dropdown_report2.library_book_dropdown_report2.update_action",
                    args: {
                        docname: book_name,
                        action: selected_value
                    },
                    callback: function() {
                        frappe.show_alert({
                            message: __("Updated action for " + book_name + " to " + selected_value),
                            indicator: "green"
                        });
                    }
                });
            });
        });
    }
};
